File: Programacao_Orientada_Objetos/Heranca/heranca.py
# Aqui na herança, você pode ver que a gente constrói novas clásses usando uma classe base
# Isso permitiria que a gente desenvolvesse classes e objetos mais complexos e úteis,
# com diferentes métodos e atributos, porém todas derivando de uma mesma base, além de
# poupar a reescrita de código

# Além disso, como você pode ver em Carro, a gente pode ir construindo encima da Classe herdada,
# colocando novos métodos e funções exclusivas a subclasse

class VeiculoAutomotor:

    def __init__(self, tipo_de_veiculo, rodas, cilindradas, portas, preco):
        self.tipo_de_veiculo = tipo_de_veiculo
        self.rodas = rodas
        self.cilindradas = cilindradas
        self.portas = portas
        self.preco = preco
        self.motor_ligado = False

    def alternar_motor(self):
        self.motor_ligado = not self.motor_ligado

    def status_veiculo(self):
        print(f"\nStatus do {self.tipo_de_veiculo}:")
        print("------------------- \n")
        print(f"Ativação do motor: {self.motor_ligado}")
        print(f"Quantidade de cilindradas: {self.cilindradas}cc")
        print(f"Quantidade de portas (não danificadas): {self.portas}")
        print(f"Quantidade de rodas: {self.rodas}")
        print(f"Preço do veículo: {self.preco}")



class Carro(VeiculoAutomotor):
    def __init__(self, rodas, cilindradas, portas, preco):
        super().__init__("Carro", rodas, cilindradas, portas, preco)
        
    def ligar_faróis(self):
        print("ligando faróis")

    def ajustar_cinto_seguranca(self):
        
        print("Cinto ajustado")

class Moto(VeiculoAutomotor):
    def __init__(self, rodas, cilindradas, portas, preco):
        super().__init__("Moto", rodas, cilindradas, portas, preco)
    
    def ajustar_cinto_seguranaca(self):
        print("Esse veículo não tem cinto")


class Onibus(VeiculoAutomotor):
    def __init__(self, rodas, cilindradas, portas, preco):
        super().__init__("Ônibus", rodas, cilindradas, portas, preco)


class Caminhao(VeiculoAutomotor):
    def __init__(self, rodas, cilindradas, portas, preco):
        super().__init__("Caminhão", rodas, cilindradas, portas, preco)


if __name__ == "__main__":
    dic = {"abc": None}
    Lamborghini = Carro(4, 1800, 4, 450000)
    Lamborghini.alternar_motor()
    Lamborghini.status_veiculo()

    Volvo = Onibus(6, 9500, 4, 395000)
    Volvo.status_veiculo()

    Kenworth = Caminhao(4, 10000, 2, 177000)
    Harley_Davidson = Moto(2, 4600, 0, 49500)

    Harley_Davidson.alternar_motor()
    Kenworth.alternar_motor()
    Kenworth.status_veiculo()

    Harley_Davidson.status_veiculo()
    Harley_Davidson.alternar_motor()
    Harley_Davidson.status_veiculo()
    
    Harley_Davidson.ajustar_cinto_seguranaca()
    Lamborghini.ajustar_cinto_seguranaca()

File: Programacao_Orientada_Objetos/Heranca/Heranca.java
package Heranca;
public class Heranca {
    public static void main(String[] args) {
        System.out.println("bananas");

        /*Veiculo_Automotor Carro = new Veiculo_Automotor("Carro", 4, 1400, 4);


        Carro.status_veiculo();
        Carro.alternar_motor();
        Carro.status_veiculo();*/

        Carro Lamborghini = new Carro(4, 1800, 4, 450000);
        Lamborghini.alternar_motor();
        Lamborghini.status_veiculo();

        Onibus Volvo = new Onibus(6, 9500, 4, 395000);
        Volvo.status_veiculo();

        Caminhao Kenworth = new Caminhao(4, 10000, 2, 177000);
        Moto Harley_Davidson = new Moto(2, 4600, 0, 49500);

        Harley_Davidson.alternar_motor();
        Kenworth.alternar_motor();
        Kenworth.status_veiculo();

        Harley_Davidson.status_veiculo();
        Harley_Davidson.alternar_motor();
        Harley_Davidson.status_veiculo();
        
    }
};

class Veiculo_Automotor {

    protected static final String CARRO = "Carro";
    protected static final String MOTO = "Moto";
    protected static final String CAMINHAO = "Caminhão";
    protected static final String ONIBUS = "Ônibus";

    final String tipo_de_veiculo;
    final int rodas;
    final int cilindradas;
    final int portas;
    private int preco;
    private boolean motor_ligado = false;

    Veiculo_Automotor(String tipo_de_veiculo, int rodas, int cilindradas, int portas, int preco) {
        
        this.tipo_de_veiculo = tipo_de_veiculo;
        this.rodas = rodas;
        this.cilindradas = cilindradas;
        this.portas = portas;
        this.preco = preco;

    };

    void alternar_motor() {
        motor_ligado = !motor_ligado;
    };

    void status_veiculo() {

        System.out.println("\nStatus do " + this.tipo_de_veiculo + ":");
        System.out.println("------------------- \n");
        System.out.println("Ativação do motor: " + this.motor_ligado);
        System.out.println("Quantidade de cilindradas: " + this.cilindradas + "cc");
        System.out.println("Quantidade de portas (não danificadas): " + this.portas);
        System.out.println("Quantidade de rodas: " + this.rodas);
        System.out.println("Preço do veículo: " + this.preco);
    };

};

class Carro extends Veiculo_Automotor {

    Carro(int rodas, int cilindradas, int portas, int preco) {
        super(CARRO, rodas, cilindradas, portas, preco);
    }

}

class Moto extends Veiculo_Automotor {

    Moto(int rodas, int cilindradas, int portas, int preco) {
        super(MOTO, rodas, cilindradas, portas, preco);
    }
}

class Onibus extends Veiculo_Automotor {

    Onibus(int rodas, int cilindradas, int portas, int preco) {
        super(ONIBUS, rodas, cilindradas, portas, preco);
    }
    
}

class Caminhao extends Veiculo_Automotor {

    Caminhao(int rodas, int cilindradas, int portas, int preco) {
        super(CAMINHAO, rodas, cilindradas, portas, preco);
    }
}